vision_doc.md

```diff
@@ -0,0 +1,279 @@
+# Vision
+`derive_builder` provides an ergonomic way to deal with construction of guaranteed-complete objects from a range of inputs:
+
+* Statically-known values
+* Function arguments
+* Wire/file/database inputs
+
+The builder should be focused specifically on the assembly of inputs and validating the completeness of the parameter set; additional validation should be handled by a constructor function provided by the target type. That constructor function may be public or private, depending on whether or not the crate author wants to _require_ the use of the builder.
+
+## Documentation
+Authors should focus documentation efforts on the target type.
+* If the builder is the only way to create the target type, then the target type should discuss the requirements for its own creation.
+* The builder can automatically include a link to the target type doc HTML from its own struct-level doc comment.
+
+## Imports
+A crate/module should always export the target type of a builder to allow the _built_ type to appear in function and struct type declarations. The exporting crate should _generally_ also expose the builder type, but may choose to exclude it from a prelude or the crate root, preferring to expose it in a child module. The crate is also free to keep the builder for its own internal use.
```

Comment:
If _only_ `FooBuilder` is exposed, then the importing crate has no way to refer to the constructed type in structs or function signatures. I can't think of a case where that's desirable.